tests/clims/models/test_substance.py

```diff
@@ -0,0 +1,197 @@
+from __future__ import absolute_import
+
+import pytest
+from django.test import TestCase
+from clims.models import Substance
+from tests.clims import testutils
+from clims.services import substances
+from django.db.models import FieldDoesNotExist
+
+
+class TestSubstance(TestCase):
+    def setUp(self):
+        self.org = testutils.create_organization()
+        self.gemstone_sample_type = testutils.create_substance_type(org=self.org)
+
+    def test_can_create_substance(self):
+        substance = substances.create(name='gem-sample-001', extensible_type=self.gemstone_sample_type, organization=self.org)
+        assert substance.version == 1
+
+    def test_can_save_substance_properties(self):
+        """
+        Saving a property through the related substance should increase the version
+        of the substance by one and set the version of the property accordingly.
+
+        It should also leave the old property unchanged
+        """
+        props = {
+            'preciousness': ':o',
+            'color': 'red'
+        }
+
+        # NOTE: One must change substances via the service rather than the django objects if all business rules
+        # are to be respected
+        substance = substances.create(
+            name='gem-sample-001',
+            extensible_type=self.gemstone_sample_type,
+            organization=self.org,
+            properties=props)
+
+        def do_asserts(substance):
+            assert substance.version == 1
+            assert len(substance.properties.all()) == 2
+            key_value_set = {(prop.extensible_property_type.name, prop.value) for prop in substance.properties.all()}
+            expected_key_value_set = {item for item in props.items()}
+            assert key_value_set == expected_key_value_set
+
+        # Assert we get the expected results on the object
+        do_asserts(substance)
+
+        # ... as well as on the object freshly fetched from the db
+        fresh_substance = Substance.objects.get(name='gem-sample-001')
+        do_asserts(fresh_substance)
+
+    def test_saving_an_unregistered_property_raises(self):
+        props = dict(weirdness=1)
+
+        with pytest.raises(FieldDoesNotExist):
+            substances.create(
+                name='gem-sample-001',
+                extensible_type=self.gemstone_sample_type,
+                organization=self.org,
+                properties=props
+            )
+
+    def test_updating_properties_via_update_or_create_updates_version(self):
+        props = dict(preciousness='*o*', color='red')
+
+        substance = substances.create(
+            name='gem-sample-001',
+            extensible_type=self.gemstone_sample_type,
+            organization=self.org,
+            properties=props)
+
+        assert substance.version == 1
```

Comment:
I dont really see the connection between substance version and substance property version in these tests. I assume substance version will increment each time substance.save is called. But what if, for example substance.save is called with the underlying substance not changed?

Reply:
Good catch! Something seems to have been lost from an earlier version of the code. I will fix this.

Reply:
... it was just the test that needed to be updated. I'm now also asserting that the substance also got an increased version.

But yes, the substance should get a new version as well as the properties that have been updated.

At the moment, I'm assuming that the name will not be updated - and if a plugin wants to allow users to update the name they'll have to have a versionable name property that can be changed(!) I consider this a potential weakness which I'd like us to look more into later. But as it stands, only the properties are updated.